### PR TITLE
[draft] Datasources: Introduce new base class

### PR DESCRIPTION
**Pull request by sunker**
**What is this feature?**

The purpose of this PR is to streamline DX and query execution flow between backend data source plugins and frontend data source plugins. 

**Why do we need this feature?**

Currently, the query execution flow for frontend and backend data source plugin differs a bit. Data sources that extend `DataSourceWithBackend` don't need to implement the query method as this is handled in the base class. The plugin just needs to implement `filterQuery` (if filtering is required) and `applyTemplateVariables`, and then these methods will be called from within the `query` method in the base class. Currently, `filterQuery` and `applyTemplateVariables` are defined in the `DataSourceWithBackend` class, although there is noting backend specific about these methods. 

Frontend plugins need to implement the query method, and handle any filtering and variable interpolation as part of that. This is causing a lot of confusion among data source developers. It would be nice if non-backend plugins could simply implement `filterQuery` and `applyTemplateVariables`, and then that would be handled by the framework before the `query` method is called. 

**How?**

This PR introduces a new base class that sits between the DataSourceAPI and DataSourceWithBackend / frontend data source plugins. This allows us to provide default implementations for common methods such as `applyTemplateVariables`. `applyTemplateVariables` and `filterQuery` are removed from DataSourceWithBackend and added to the query runner, enabling the same query execution flow for frontend and backend plugins. 

datasourcebase

This change would have no impact on backend data sources. For frontend plugins. things would also work as before. However, from G11 we would encourage them to extend `DataSourceBase` rather than `DataSourceAPI`, and simply implement `applyTemplateVariables` and `filterQuery`. 

@ryantxu do you know if this comment below is still valid? If not, we can skip DataSourceBase and just implement defaults in DataSourceAPI? But I don't know - maybe useful to have this base class anyway? Feels like a lot of the props and methods (´variables´, ´annotations´, ´targetContainsTemplate´) could make use of decent default implementation. 
https://github.com/grafana/grafana/blob/acf97e43b664cf966f4c85d53580e4a5573e59cc/packages/grafana-data/src/types/datasource.ts#L194-L195

**Who is this feature for?**

Plugin developers

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

This is just a draft PR, so I have not yet updated the query runner for alerting, annotations etc. 

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-runtime/src/utils/DataSourceBase.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  /**
+   * Override to apply template variables and adhoc filters.  The result is usually also `TQuery`, but sometimes this can
+   * be used to modify the query structure before sending to the backend.
+   *
+   * NOTE: if you do modify the structure or use template variables, alerting queries may not work
+   * as expected
+   *
+   * @virtual
+   */
+  applyTemplateVariables(query: TQuery, scopedVars: ScopedVars, filters?: AdHocVariableFilter[]) {
```
Comment (sunker) [inline packages/grafana-runtime/src/utils/DataSourceBase.ts:58]: Would be nice to make this abstract as all data sources should interpolate variables, but maybe a bit too brave?

File: packages/grafana-runtime/src/utils/DataSourceBase.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   * Query for data, and optionally stream results
+   */
+  abstract query(request: DataQueryRequest<TQuery>): Promise<DataQueryResponse> | Observable<DataQueryResponse>;
+
+  /**
+   * Override to apply template variables and adhoc filters.  The result is usually also `TQuery`, but sometimes this can
+   * be used to modify the query structure before sending to the backend.
+   *
+   */
+  applyTemplateVariables(query: TQuery, scopedVars: ScopedVars, filters?: AdHocVariableFilter[]): TQuery {
```
Comment (sunker) [inline packages/grafana-runtime/src/utils/DataSourceBase.ts:47]: All plugins _should_ implement this method, so would be nice to make it abstract.

File: packages/grafana-runtime/src/utils/DataSourceBase.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+} from '@grafana/data';
+import { DataQuery, DataSourceJsonData } from '@grafana/schema';
+
+/**
+ * This class provides generic default implementations for data source plugin methods.
+ * Extend this class to implement a data source plugin.
+ *
+ * @public
+ */
+export abstract class DataSourceBase<
```
Comment (mckn) [inline packages/grafana-runtime/src/utils/DataSourceBase.ts:20]: Maybe move this to `grafana/data` to be placed togheter with the other ds types.

**Review by mckn — APPROVED** [review]
I think this looks good! We should probably add something to the migration guide for plugins:

https://grafana.com/developers/plugin-tools/migration-guides/update-from-grafana-versions/

**Comment by sunker** [comment]
Closing in favour of https://github.com/grafana/grafana/pull/84656

**Pull request closed without merging**